OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs

```diff
@@ -0,0 +1,131 @@
+	class PaletteFromGimpOrJascFileInfo : ITraitInfo
+	{
+		[FieldLoader.Require, PaletteDefinition]
+		[Desc("Palette name used internally.")]
+		public readonly string Name = null;
+
+		[Desc("Defines for which tileset IDs this palette should be loaded.",
+			"If none specified, it applies to all tileset IDs not explicitly excluded.")]
+		public readonly HashSet<string> Tilesets = new HashSet<string>();
+
+		[Desc("Don't load palette for these tileset IDs.")]
+		public readonly HashSet<string> ExcludeTilesets = new HashSet<string>();
+
+		[FieldLoader.Require]
+		[Desc("Name of the file to load.")]
+		public readonly string Filename = null;
+
+		[Desc("Map listed indices to shadow. Ignores previous color.")]
+		public readonly int[] ShadowIndex = { };
+
+		[Desc("Premultiply colors with their alpha values.")]
+		public readonly bool Premultiply = true;
+
+		public readonly bool AllowModifiers = true;
+
+		public object Create(ActorInitializer init) { return new PaletteFromGimpOrJascFile(init.World, this); }
+	}
+
+	class PaletteFromGimpOrJascFile : ILoadsPalettes, IProvidesAssetBrowserPalettes
+	{
+		readonly World world;
+		readonly PaletteFromGimpOrJascFileInfo info;
+
+		public PaletteFromGimpOrJascFile(World world, PaletteFromGimpOrJascFileInfo info)
+		{
+			this.world = world;
+			this.info = info;
+		}
+
+		public void LoadPalettes(WorldRenderer wr)
+		{
+			var colors = new uint[Palette.Size];
+			using (var s = world.Map.Open(info.Filename))
+			{
+				using (var lines = s.ReadAllLines().GetEnumerator())
+				{
+					if (lines == null)
+						return;
+
+					if (!lines.MoveNext() || (lines.Current != "GIMP Palette" && lines.Current != "JASC-PAL"))
+						throw new InvalidDataException("File `{0}` is not a valid GIMP or JASC palette.".F(info.Filename));
+
+					byte r, g, b, a;
+					a = 0;
+					var i = 0;
+
+					while (lines.MoveNext() && i < Palette.Size)
+					{
+						// Skip until first color. Ignore # comments, Name/Columns and blank lines as well as JASC header values.
+						if (string.IsNullOrEmpty(lines.Current) || !char.IsDigit(lines.Current.Trim()[0]) || lines.Current == "0100" || lines.Current == "256")
+							continue;
+
+						var rgba = lines.Current.Split((char[])null, StringSplitOptions.RemoveEmptyEntries);
+						if (rgba.Length < 3)
+							throw new InvalidDataException("Invalid RGB(A) triplet/quartet: ({0})".F(string.Join(" ", rgba)));
+
+						if (!byte.TryParse(rgba[0], out r))
+							throw new InvalidDataException("Invalid R value: {0}".F(rgba[0]));
+
+						if (!byte.TryParse(rgba[1], out g))
+							throw new InvalidDataException("Invalid G value: {0}".F(rgba[1]));
+
+						if (!byte.TryParse(rgba[2], out b))
+							throw new InvalidDataException("Invalid B value: {0}".F(rgba[2]));
+
+						if (rgba.Length == 3 || (rgba.Length > 3 && !byte.TryParse(rgba[3], out a)))
+							a = 0;
+
+						if (r == 0 && g == 0 && b == 0)
+							colors[i] = 0;
+						else if (a == 0)
```

Comment:
Doesn't this prevent fully transparent colours?

Reply:
The issues here could be solved by treating `r == 0 && g == 0 && b == 0 && a == 0` (transparent) as distinct from `r == 0 && g == 0 && b == 0` (black)

Reply:
But alpha of 0 means zero transparency...?

Reply:
Sorry yes, I meant non-transparent black on the case above.